module/VuFindSearch/src/VuFindSearch/Backend/Blender/Response/Json/RecordCollection.php

```diff
@@ -0,0 +1,441 @@
+                $this->add(array_shift($backendRecords[$backendId]), false);
+            }
+        }
+
+        $this->response['facet_counts']['facet_fields']
+            = $this->getMergedFacets($collections);
+
+        return $backendRecords;
+    }
+
+    /**
+     * Slice the record collection
+     *
+     * @param int $offset Offset
+     * @param int $limit  Limit
+     *
+     * @return void
+     */
+    public function slice(int $offset, int $limit): void
+    {
+        $this->records = array_slice(
+            $this->records,
+            $offset,
+            $limit
+        );
+    }
+
+    /**
+     * Add an error message
+     *
+     * @param mixed $error Error
+     *
+     * @return void
+     */
+    public function addError($error): void
+    {
+        if (!in_array($error, $this->errors)) {
+            $this->errors[] = $error;
+        }
+    }
+
+    /**
+     * Return any errors.
+     *
+     * Each error can be a translatable string or an array that the Flashmessages
+     * view helper understands.
+     *
+     * @return array
+     */
+    public function getErrors()
+    {
+        return $this->errors;
+    }
+
+    /**
+     * Set the source backend identifier.
+     *
+     * @param string $identifier Backend identifier
+     *
+     * @return void
+     */
+    public function setSourceIdentifier($identifier)
+    {
+        $this->source = $identifier;
+    }
+
+    /**
+     * Calculate the backend to be used for the record at a given position
+     *
+     * Note: This does not take into account whether there are enough records in the
+     * source.
+     *
+     * @param int   $position       Position
+     * @param int   $blockSize      Record block size
+     * @param array $backendIds     Available backends
+     * @param array $initialResults List of backends for initial result boosts
+     *
+     * @return string
+     */
+    protected function getBackendAtPosition(
+        int $position,
+        int $blockSize,
+        array $backendIds,
+        array $initialResults
+    ): string {
+        if ($boostBackend = $initialResults[$position] ?? false) {
+            return $boostBackend;
+        }
+
+        // We're outside the blocks affected by boosting, calculate by block
+        $currentBlock = floor($position / $blockSize);
+        $backendCount = count($backendIds);
+        return $backendCount ? $backendIds[$currentBlock % $backendCount] : '';
+    }
+
+    /**
+     * Merge facets
+     *
+     * @param array $collections Result collections
+     *
+     * @return array
+     */
+    protected function getMergedFacets(array $collections): array
+    {
+        $mergedFacets = [];
+
+        // Iterate through mappings and merge values. It is important to do it this
+        // way since multiple facets may to a single one.
+        foreach ($this->mappings['Facets']['Fields'] ?? []
+            as $facetField => $settings
+        ) {
+            $list = [];
```

Comment:
Might it make sense to refactor all the code that populates $list to a support function for better readability?

Reply:
Refactored.